### ISSUE TITLE
Linux Flatpak / AppImage releases

### ISSUE THREAD

**Issue by AllienWorks** [sparrowwallet/sparrow#1]
First of all, based on the site, Sparrow looks really cool! Can't wait to test it out.

However, not all Linux distros support DEB/RPM packages – do you have any plans to releasing also via [Flatpak](https://flatpak.org/) or [AppImage](https://appimage.org/) (which are both quite universal Linux packages)?

**Comment by monperrus** [comment]
Yes, I also need to run sparrow on a machine on which I'm not root.

A standalone build of sparrow for Linux would be awesome.

**Comment by craigraw** [comment]
Sparrow v0.9.8 has been released with a Linux standalone build. It's not a Flatpak or AppImage, but just a tar.gz of the binary and associated libraries. 

Does this work?

**Comment by monperrus** [comment]
Yes, it works, thanks a lot. Had to `chmod 755 bin/Sparrow` first.

**Comment by AllienWorks** [comment]
Yep, I can confirm. After chmodding the binary, it runs fine.
Thanks much @craigraw!

**Comment by craigraw** [comment]
Thanks for the feedback! I'll still look into an AppImage in future.

**Comment by 6102bitcoin** [comment]
Outstanding Action: Add AppImage release
Proposed Priority: Low

**Comment by mariaa144** [comment]
> Sparrow v0.9.8 has been released with a Linux standalone build. It's not a Flatpak or AppImage, but just a tar.gz of the binary and associated libraries.
> 
> Does this work?

The binary doesn't work for NixOS users. NixOS does have a package for Sparrow but it is currently broken. An Appimage or Flatpak would be a nice addition.

**Comment by ademar111190** [comment]
There is an official [SDK Extension for OpenJDK 20](https://github.com/flathub/org.freedesktop.Sdk.Extension.openjdk) on flathub repo; I believe that can help a flatpak support effort.

**Comment by craigraw** [comment]
Thanks - Sparrow already uses `jlink` to package a minimized JRE, so this shouldn't be necessary. It's effectively a binary application.

**Comment by storopoli** [comment]
Is this legit?
It has the verified icon in Flathub: https://flathub.org/apps/io.github.Strozer2002.sparrow

**Comment by RequestPrivacy** [comment]
> Is this legit?

I don't think so, the question is can something be done against this?

**Comment by monperrus** [comment]
a compromised version through a package manager? (remember the Ledger attack on the Microsoft store 
https://www.hackread.com/fake-ledger-app-microsoft-app-store-crypto-funds/)

**Comment by craigraw** [comment]
Definitely not legit, and probably malware.

The download size, version number and app architecture (appears to be a flutter app) are all dramatically different.

I have reported to flathub@lists.freedesktop.org. I would appreciate any help in getting this removed.

**Comment by craigraw** [comment]
Have also reported the actual application at https://github.com/Strozer2002/sparrow-linux/ to Github.

**Comment by xmready** [comment]
@craigraw I would seriously recommend having a verified flatpak version on flathub. There are multiple pros for this:

1. Reduces attack vector of non-official flatpaks
2. Flatpaks can update automatically removing the need to manually update Sparrow
3. Flatpaks work across all popular Linux distros
4. Flatpaks have become extremely popular
5. It's less complicated than changing permissions on a standalone binary for noobs

**Comment by craigraw** [comment]
@xmready The problem with supporting flathub (or another app store like snap store, winget or the mac app store) is that it broadens the distribution attack vector. Right now, the only source for Sparrow binaries is Github. These binaries are independently watched by https://binarywatch.org. The security model for app stores is generally more complicated, and they don't usually have robust independent monitoring of binaries afaik.

**Comment by xmready** [comment]
> @xmready The problem with supporting flathub (or another app store like snap store, winget or the mac app store) is that it broadens the distribution attack vector. Right now, the only source for Sparrow binaries is Github. These binaries are independently watched by https://binarywatch.org. The security model for app stores is generally more complicated, and they don't usually have robust independent monitoring of binaries afaik.

My understanding is that a verified Flatpak submission on Flathub comes from a GitHub PR made by the original developer, aka you. The Flatpak is built using your project's repo. Are you saying that the resulting build can't be verified?

**Comment by ziqiq** [comment]
I created [a Flatpak manifest file](https://gist.github.com/ziqiq/9498f030e503d59a10fdc9afc01a2a6a) and installed it on Fedora 40/GNOME/X11

How to reproduce:

1. Install build dependency `flatpak-builder`
```
dnf install flatpak-builder
```
2. Build (and install) the app
```
flatpak-builder --force-clean --user --install-deps-from=flathub --repo=repo --install builddir com.sparrowwallet.Sparrow.yml
```
This will create two directories:
  - `builddir`, the actual files built for the Flatpak app
  - `repo`, a local Flatpak repo where the app published to
3. If a bundle file is needed for distribution
```
flatpak build-bundle repo sparrow-1.9.1-x86_64.flatpak com.sparrowwallet.Sparrow 1.9.1 --runtime-repo=https://flathub.org/repo/flathub.flatpakrepo
```
This will create a single-file Flatpak bundle `sparrow-1.9.1-x86_64.flatpak` 

See [the Flatpak build guide](https://docs.flatpak.org/en/latest/first-build.html) for detail.

Note:
- The Sparrow home directory is `~/.var/app/com.sparrowwallet.Sparrow/.sparrow`. You might want to copy the existing data file from `~/.sparrow` into it (or create a symbolic link). 
- As I added a desktop file, there will be an Sparrow entry in the launcher:)

**Comment by ziqiq** [comment]
Some other FYIs:

1. I just realized the `flatpak-builder` itself is also available on Flathub as a Flatpak app, so if you are not comfortable with installing additional rpm/deb package to build the manifest, or don't have root permission, you could:

```
flatpak install --user org.flatpak.Builder
alias flatpak-builder='flatpak run org.flatpak.Builder'
```

2. If you don't need the bundle file, you could omit `--repo=repo`. Also, usually most environment should have the base Freedesktop runtime, so I think most end user might just run this to install Sparrow as Flatpak:

```
flatpak-builder --user --install builddir com.sparrowwallet.Sparrow.yml
```

**Comment by craigraw** [comment]
> Are you saying that the resulting build can't be verified?

@xmready No, please read my reply again.

**Comment by jwillikers** [comment]
> > Are you saying that the resulting build can't be verified?
> 
> @xmready No, please read my reply again.

To summarize, the problems with app stores are:

1. They add an additional node in the supply chain, which increases the attack surface.
2. There isn't robust, independent monitoring for binaries provided from app stores.

Did I get that right?

The first point will hold for distribution on any third-party app store, while the second point may eventually be addressed.

In regards to Flatpak distribution through a remote, these points definitely still hold.
Even if the Flatpak is provided from your own remote, it's still another node in the graph not to mention the maintenance burden, even if it's simply hosted on GitHub Pages.
Point two would still be problematic, since it's not just a tarball being provided, but really binaries / binary deltas via ostree.

It is possible to distribute Flatpaks without a remote, which would address both points.
Unfortunately, this loses one of the major benefits a lot users probably want: automatic updates.

It's also worth pointing out that Flatpak runtimes are managed independent of the application, creating another attack vector. While this is somewhat analogous to the attack vector that is your OS or distribution, it is an additional layer.

**Comment by craigraw** [comment]
@jwillikers Yes that sounds correct. Thanks for the additional context.

One thing I'll add is that you generally don't want automatic updates when it comes to Bitcoin software, as they again increase the attack surface.

**Comment by xmready** [comment]
@craigraw would you ever consider hosting your own .deb repository as an alternative?

**Comment by rdong8** [comment]
With Snap packages ([`bitcoin-core` is officially distributed through snap](https://snapcraft.io/bitcoin-core)) you can fix the version by appending ie. `--channel=27.1/stable` to the `snap install` command, so that's another option.

**Comment by DeepDoge** [comment]
AppImage would work great, things like "Gear Lever" makes it an amazing experience, even on atomic distributions.

**Comment by davideb263** [comment]
> AppImage would work great, things like "Gear Lever" makes it an amazing experience, even on atomic distributions.

II second this. Gear Lever makes it very easy to manage updates by simply linking the app to the GitHub releases. It's not completely automatic but you just need to click a notification, so I think it could be a good compromise.

**Comment by DeepDoge** [comment]
I just switched to using distrobox and exporting the launcher to the host.
You dont even need to open the terminal.

- Make sure you have Distrobox, I'm using Fedora Silverblue with `ghcr.io/ublue-os/silverblue-nvidia` image, and it has it built-in already.
- Install and start BoxBuddy flatpak app.
- Create a distrobox with the image `ubuntu-toolbox`.
- - I named the distrobox "bitcoin". because i install other bitcoin related stuff in it as well.
- - You might wanna toggle-on "Use init system", when not enabled, causes issues with systemd services. If you ever wanna install something with a service later in the same distrobox. 
- - Optionally If you don't like home folder littering, you can give your distrobox a different HOME folder, I used `~/Distrobox/sparrow`.
- Download `.deb` package.
- In the BoxBuddy select "Install .deb file", select the file and it will install it.
- After installation, Click on "View Applications" on BoxBuddy, and click on "Add To Menu" for "Sparrow".

Of course you can do these all with cli too.

**Comment by skobkin** [comment]
Just my two cents here.

I think that not distributing the app for those who need it in a convenient way is also a risk in reality.

I'm not saying about supporting each package manager in the world, but systems like Flatpak provide a reasonable compromise between developer's efforts and user's convenience and safety.

A *lot of people* appreciate *convenience*. And someday (if not already) users will start to use some 3rd party Flatpak repo or even 3rd-party package on Flathub.

Yes, if something happens the project can always state that it was a 3rd-party distribution. But that doesn't make users more secure, it just telegraphs a message to them: "sorry, but it was your fault".

At the same time I totally can understand why a wallet vendor doesn't want to have additional potentially vulnerable link in the distribution chain.

But from the viewpoint of how life usually works, we all know that in the end users will run shady `curl | sh` one-liners from internet, install 3rd-party Flatpaks and so on.

If it'd be me, I'd choose a compromise:

- provide builds which are convenient for users
- leave a warning that technically those 3rd-party app stores are less secure than using binaries from official Github repo

P.S. Yes, I understand that the "sending a message" part sounds a bit manipulative and I'm sorry for that, but it's also partially true from my PoV.

**Comment by Sami32** [comment]
FYI:
https://blog.sebastianwick.net/posts/flatpak-happenings/

I admit that it would be nice to be able to use a such Sparrow package with TailsOS by the end of 2027 :-)
